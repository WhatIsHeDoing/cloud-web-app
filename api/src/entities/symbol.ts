import { Column, Entity, ObjectID, ObjectIdColumn } from "typeorm";

/**
 * A cleaner representation of a symbol from an Alpha Vantage search.
 * @see https://www.alphavantage.co/documentation/#symbolsearch
 */
@Entity()
export class Symbol {
    @Column()
    currency: string;

    /**
     * Although the symbol is technically the ID, TypeORM requires a UUID.
     * @see https://github.com/typeorm/typeorm/issues/2530
     */
    @ObjectIdColumn()
    id: ObjectID;

    @Column()
    marketClose: string;

    @Column()
    marketOpen: string;

    @Column()
    name: string;

    @Column()
    region: string;

    @Column()
    symbol: string;

    @Column()
    timezone: string;

    @Column()
    type: string;
}
