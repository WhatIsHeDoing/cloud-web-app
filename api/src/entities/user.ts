import { Column, Entity, ObjectID, ObjectIdColumn } from "typeorm";

@Entity()
export class User {
    @ObjectIdColumn()
    id: ObjectID;

    @Column()
    firstName: string;

    @Column()
    lastName: string;
}
