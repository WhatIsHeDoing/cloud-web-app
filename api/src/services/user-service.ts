import { PubSub } from "apollo-server";
import { getMongoManager } from "typeorm";
import { User } from "../entities";
import { CreateUserOptions, SubscriptionUpdateType, UserPatch } from "../models";

export const createUser = async (opts: CreateUserOptions): Promise<User> => {
    const manager = getMongoManager();
    const user = manager.create(User, opts);
    await manager.save(user);
    await publishChange("CREATE", user);
    return user;
};

export const deleteUser = async (id: string): Promise<User | null> => {
    const manager = getMongoManager();
    const user = await manager.findOne(User, id);

    if (!user) {
        return null;
    }

    const userClone = { ...user };
    await manager.remove(user);
    await publishChange("DELETE", user);
    return userClone;
};

export const deleteUsers = async (): Promise<User[]> => {
    const manager = getMongoManager();
    const allUsers = await manager.find(User);
    await manager.deleteMany(User, {});
    await Promise.all(allUsers.map(user => publishChange("DELETE", user)));
    return allUsers;
};

export const findAllUsers = (): Promise<User[]> => getMongoManager().find(User);

export const subscribeToUserChanges = () => _pubsub.asyncIterator(CHANNEL_NAME);

export const updateUser = async (id: string, patch: UserPatch): Promise<User | null> => {
    const manager = getMongoManager();
    const user = await manager.findOne(User, id);

    if (!user) {
        return null;
    }

    Object.assign(user, patch);
    await manager.save(user);
    await publishChange("UPDATE", user);
    return user;
};

const _pubsub = new PubSub();

const CHANNEL_NAME = "USER_ACTION";

const publishChange = (updateType: SubscriptionUpdateType, user: User) => _pubsub.publish(
    CHANNEL_NAME,
    {
        userUpdated: {
            updateType,
            user
        }
    });
