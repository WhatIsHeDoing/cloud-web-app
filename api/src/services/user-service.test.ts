import * as typeorm from "typeorm";
import { User } from "../entities";
import { findAllUsers } from "./user-service";

describe("user-service", () => {
    describe("findAllUsers", () => {
        it("handles some users", () => {
            const mockGetMongoManager = ({
                find: async (_: any) => new Promise<User[]>((resolve, _reject) =>
                    resolve([
                        {
                            firstName: "Amanda",
                            lastName: "Huggenkiss"
                        } as User
                    ])
                )
            } as typeorm.MongoEntityManager);

            const mock = jest.spyOn(typeorm, "getMongoManager");
            mock.mockReturnValue(mockGetMongoManager);

            // Leave Jest to resolve the promise.
            // tslint:disable-next-line: no-floating-promises
            expect(findAllUsers()).resolves.toMatchSnapshot();
        });

        it("handles no users", () => {
            const mockGetMongoManager = ({
                find: async (_: any) => new Promise<User[]>((resolve, _reject) => resolve([]))
            } as typeorm.MongoEntityManager);

            const mock = jest.spyOn(typeorm, "getMongoManager");
            mock.mockReturnValue(mockGetMongoManager);

            // Leave Jest to resolve the promise.
            // tslint:disable-next-line: no-floating-promises
            expect(findAllUsers()).resolves.toHaveLength(0);
        });
    });
});
