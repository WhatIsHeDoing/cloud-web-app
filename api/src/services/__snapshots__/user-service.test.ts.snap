// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user-service findAllUsers handles some users 1`] = `
Array [
  Object {
    "firstName": "Amanda",
    "lastName": "Huggenkiss",
  },
]
`;
